src/connstat.cpp

```diff
@@ -0,0 +1,175 @@
+            if (((ConnectivityStatus*)data)->logger_)
+                ((ConnectivityStatus*)data)->logger_->w("taking care of ACK\n");
+        return (((ConnectivityStatus*)data)->stop) ? NL_OK : NL_STOP;
+    }, (void*)this);
+
+
+    nl_send_auto(nlsk.get(), bye.get());
+
+    if (thrd_.joinable())
+        thrd_.join();
+
+}
+
+void
+ConnectivityStatus::setEventListener(ConnectionEventCb ucb, Event event)
+{
+    std::lock_guard<std::mutex> lck(mtx_);
+    switch (event) {
+    case Event::NEWADDR:
+    case Event::DELADDR:
+    case Event::ADDR:
+        nl_socket_add_memberships(nlsk.get(), RTNLGRP_IPV6_IFADDR, RTNLGRP_NONE);
+        nl_socket_add_memberships(nlsk.get(), RTNLGRP_IPV4_IFADDR, RTNLGRP_NONE);
+        break;
+    default:
+        break;
+    }
+    event_cbs[event] = ucb;
+}
+
+void
+ConnectivityStatus::removeEventListener(Event event)
+{
+    std::lock_guard<std::mutex> lck(mtx_);
+    event_cbs.erase(event);
+    switch (event) {
+    case Event::NEWADDR:
+    case Event::DELADDR:
+    case Event::ADDR:
+        nl_socket_drop_memberships(nlsk.get(), RTNLGRP_IPV6_IFADDR, RTNLGRP_NONE);
+        nl_socket_drop_memberships(nlsk.get(), RTNLGRP_IPV4_IFADDR, RTNLGRP_NONE);
+        break;
+    default:
+        break;
+    }
+}
+
+void
+ConnectivityStatus::executer(Event event)
+{
+    auto cb = event_cbs.find(event);
+    if (cb != event_cbs.end() && cb->second)
+        (cb->second)(event);
+}
+
+
+void
+ConnectivityStatus::nlsk_setup(nl_sock* nlsk)
+{
+    nl_socket_disable_seq_check(nlsk);
+    nl_socket_set_nonblocking(nlsk);
+
+    nl_socket_modify_cb(nlsk, NL_CB_VALID, NL_CB_CUSTOM, [](nl_msg* msg, void* data) -> int {
+            return ((ConnectivityStatus*)data)->nl_event_cb(msg);
+    }, (void*)this);
+
+    nl_connect(nlsk, NETLINK_ROUTE);
+}
+
+ConnectivityStatus::NlPtr
+ConnectivityStatus::nlsk_init(void)
+{
+    NlPtr ret(nl_socket_alloc(), &nl_socket_free);
+    if (not ret.get())
+        throw std::runtime_error("couldn't allocate netlink socket!\n");
+    else
+        return ret;
+}
+
+void
+ConnectivityStatus::get_neigh_state(struct nl_msg* msg)
+{
+    struct nlmsghdr *h = nlmsg_hdr(msg);
+    struct ndmsg *r = (struct ndmsg*)nlmsg_data(h);
+
+    switch (r->ndm_state) {
+    case NUD_REACHABLE:
+        executer(Event::NEWNEIGH);
+        break;
+    default:
+        break;
+    }
+}
+
+int
+ConnectivityStatus::nl_event_cb(struct nl_msg* msg)
+{
+    std::lock_guard<std::mutex> lck(mtx_);
+    struct nlmsghdr *h = nlmsg_hdr(msg);
+
+    int status = NL_OK;
+    switch (h->nlmsg_type) {
+    case RTM_NEWADDR:
+        executer(Event::ADDR);
+        executer(Event::NEWADDR);
+        break;
+    case RTM_DELADDR:
+        executer(Event::ADDR);
+        executer(Event::DELADDR);
+        break;
+    case RTM_NEWNEIGH:
+    case RTM_DELNEIGH:
+    case RTM_NEWNEIGHTBL:
+        get_neigh_state(msg);
+        break;
+    default:
+        status = NL_SKIP;
+        break;
+    }
+    return status;
```

Comment:
simply return NL_STOP if stopping here ?

Reply:
```NL_STOP``` actually reports error; however for ```NL_CB_VALID``` it wouldn't make a difference.

i moved us to ```NL_CB_MSG_IN``` which means more control; e.g. terminating early since we get to stop early and possibly exit before libnl starts allocate things for it's routines; ack reply for instance.
anyway, this means that we should return ```NL_OK``` so that nllib continues with its routine in case we're not terminating; continuing with next msg for multipart ones or sending back ack to kernel and so on.

a side note, the changes i'm about to push to this are not testing on termination because of the problem we saw the other day.